src/DynamoCore/Graph/ModelBase.cs

```diff
@@ -211,6 +212,21 @@ public bool IsSelected
             }
         }
 
+        /// <summary>
+        /// If this model belongs to an AnnotationModel
+        /// </summary>
+        [JsonIgnore]
+        public bool BelongsToGroup
```

Comment:
@mjkkirschner @aparajit-pratap What do you feel about adding this property to `ModelBase`?

Reply:
It seems a bit odd - groups already have a list of models that they contain, and not ALL models can belong to a group - ie connectors.

Reply:
Fair enough, the reason i felt this was needed was that I have to do this check a lot for the drag and drop functionality. Essentially i need to check if the dragged model belongs to a group every time its dragged over a group, it seems very inefficient to have to go over all Groups and check its nodes collection everytime this happens.

I get that there are some Models that does not have any functionality that enables it to be added to a group, but at the same time the `Nodes` collection is of type `IEnumerable<ModelBase>` which does indicate that in theory, it can take all `ModelBase`s.

Happy to revert this and simply do the `AnnotaitonModel.Nodes.Contains(modelBase)` check every time, my worry was just on performance as this has to be done a lot.

https://github.com/DynamoDS/Dynamo/pull/11912/files#diff-32e4994e829fb0abe599845f0677a235d127cf860b8a85eb31667dfef50db6c0R822

Reply:
@SHKnudsen if you are worried about the performance one option is replace the implementation in AnnotationModel with a hashset of nodes. Should there ever be duplicates in the group?

Reply:
No there won't ever be duplicates, my worry was more that I need to go through every group in the graph here to check if the dragged node belongs to any of them, so if there was a large number of groups in the graph it seemed inefficient to go through them all every time i need to do this check.

the workflow that has to happen when a node is repositioned: 
- check if its over a group (we have to use the mouse cursor position here)
    - because we are dragging another element we cant use `MouseOver` event on the group
- then check if the dragged node already belongs to a group (this is where it was convenient to simply do `draggedNode.BelongsToGroup`)
 - if we are hovering over a group and the node does not already belong to a group then set the group in a `NodeHovering state` so the group ui gets a grey border around it indicating that we can drop the node in it

Reply:
okay thanks for the detailed description @SHKnudsen - 

1. definitely sounds like it could get slow at the moment you need to check what group you are over if you need to iterate every group, and every node in that group using a list of nodes (how it's implemented today). My suggestion was to modify the `AnnotationModel.Nodes` internal datastructure to be a hashSet or dictionary so that the lookup time is constant for each group, then you just need to iterate over all groups - a much smaller collection.

2. Checking the mouse position is fraught - it will need to be tested with different DPI scaling settings applied.
3. If you feel your solution is better (personally I don't like trying to keep state like that correct) - then what about an interface that NodeModel and AnnotationModel can implement. `ICanBelongToGroup` ?

Reply:
I agree with Mike that iterating over existing groups only when MouseUp is triggered should not cause that much of a performance penalty. 
One other solution could be a new static (or Singleton ) class that can store a map of NodeId to GroupId, This data can be populated when Dynamo starts up and when any group is altered.